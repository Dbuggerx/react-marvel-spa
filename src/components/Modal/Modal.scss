@import '../scss/variables';
@import '../scss/placeholders';

.modal {
  background: rgba(53, 24, 22, 0.8);
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  flex-flow: column nowrap;
  display: none;
  z-index: 10;

  &--opened {
    display: flex;
    flex-flow: column nowrap;
    justify-content: center;
    align-items: stretch;
  }

  &__container {
    flex: 1 1 auto;
    margin: 5vh 15vw;
    background: map-get($colors, highlight-bg);
    color: map-get($colors, highlight-fg);
    display: flex;
    flex-flow: column nowrap;
    position: relative;
    overflow: auto;
    border: 0;
    padding: 15px;
    width: unset;
    height: unset;

    @media #{map-get($breakpoints, small)} {
      margin: 5vh;
    }
  }

  &__close {
    @extend %unstyled-button;
    position: absolute;
    right: 15px;
    top: 15px;    
    font-size: 2em;
    line-height: 1em;
  }

  &__title {
    line-height: 2em;
    font-size: 2em;
    text-align: center;
    color: map-get($colors, marvel-red);
  }
}
