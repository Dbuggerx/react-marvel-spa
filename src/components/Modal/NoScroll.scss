.no-scroll {
  overflow: hidden;
  max-height: 100vh;
}
