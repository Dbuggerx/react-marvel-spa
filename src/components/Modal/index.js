// @flow
import React, { Component } from 'react';
import './Modal.scss';
import './NoScroll.scss';

type Props = {
    children: any,
    title: string,
    onClose: () => void,
};

export default class Modal extends Component<Props> {
    constructor(props: Props) {
        super(props);
        this.closeButton = React.createRef();
        this.container = React.createRef();
    }

    componentDidMount() {
        this.previouslyActiveElement = document.activeElement;
        if (this.closeButton.current) this.closeButton.current.focus();
        window.addEventListener('keyup', this.handleCloseKeyUp);
        window.addEventListener('focus', this.handleFocus, true);
        if (document.body) document.body.classList.add('no-scroll');
    }

    componentWillUnmount() {
        window.removeEventListener('keyup', this.handleCloseKeyUp);
        window.removeEventListener('focus', this.handleFocus, true);
        if (document.body) document.body.classList.remove('no-scroll');
        if (this.previouslyActiveElement) this.previouslyActiveElement.focus();
    }

    closeButton: *;
    container: *;
    previouslyActiveElement: ?HTMLElement;

    handleCloseKeyUp = (event: SyntheticKeyboardEvent<*>) => {
        const escapeKey = 'key' in event ? event.key === 'Escape' || event.key === 'Esc' : event.keyCode === 27;
        if (escapeKey) this.props.onClose();
    };

    handleFocus = (event: SyntheticEvent<*>) => {
        if (
            event.target !== event.currentTarget &&
            this.container.current &&
            this.closeButton.current &&
            !this.container.current.contains(event.target)
        ) {
            this.closeButton.current.focus();
        }
    };

    render() {
        return (
            <div className="modal modal--opened">
                <dialog
                    open
                    className="modal__container"
                    aria-modal="true"
                    aria-labelledby="modal-title"
                    aria-describedby="modal-content"
                    ref={this.container}
                >
                    <button
                        type="button"
                        className="modal__close"
                        aria-label="Close"
                        onKeyUp={this.handleCloseKeyUp}
                        onClick={this.props.onClose}
                        tabIndex="0"
                        ref={this.closeButton}
                    >
                        <span aria-hidden="true">&times;</span>
                    </button>
                    <div className="modal__title" id="modal-title">
                        {this.props.title}
                    </div>
                    <div className="modal__content" id="modal-content">
                        {this.props.children}
                    </div>
                </dialog>
            </div>
        );
    }
}
