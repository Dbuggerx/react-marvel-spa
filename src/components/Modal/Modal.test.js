import React from 'react';
import { shallow } from 'enzyme';
import Modal from './index';

describe('Modal', () => {
    describe('render', () => {
        test('renders properly', () => {
            const wrapper = shallow(<Modal title="Testing">Test</Modal>);
            expect(wrapper).toMatchSnapshot();
        });

        test('close button calls "props.onClose" upon click', () => {
            const onCloseSpy = jest.fn();
            const wrapper = shallow(<Modal onClose={onCloseSpy} />);
            wrapper.find('.modal__close').simulate('click');

            expect(onCloseSpy).toHaveBeenCalledTimes(1);
        });

        test('close button calls "props.onClose" upon Esc keyUp', () => {
            const onCloseSpy = jest.fn();
            const wrapper = shallow(<Modal onClose={onCloseSpy} />);
            const closeButon = wrapper.find('.modal__close');
            closeButon.simulate('keyUp', {
                key: 'Esc',
            });
            closeButon.simulate('keyUp', {
                keyCode: 27,
            });
            closeButon.simulate('keyUp', {
                key: 'wrongOne',
            });
            expect(onCloseSpy).toHaveBeenCalledTimes(2);
        });
    });

    describe('componentDidMount', () => {
        test('focuses on the close button', () => {
            const wrapper = shallow(<Modal />);
            const focusSpy = jest.fn();

            wrapper.instance().closeButton = {
                current: {
                    focus: focusSpy,
                },
            };
            wrapper.instance().componentDidMount();
            expect(focusSpy).toHaveBeenCalledTimes(1);
        });

        test('adds event listeners to window', () => {
            window.addEventListener = jest.fn();
            const wrapper = shallow(<Modal />);
            expect(window.addEventListener).toHaveBeenCalledTimes(2);
            expect(window.addEventListener).toHaveBeenCalledWith('keyup', wrapper.instance().handleCloseKeyUp);
            expect(window.addEventListener).toHaveBeenCalledWith('focus', wrapper.instance().handleFocus, true);
        });

        test('adds "no-scroll" class to body', () => {
            document.body.classList.remove('no-scroll');
            expect(document.body.classList).not.toContain('no-scroll');
            shallow(<Modal />);
            expect(document.body.classList).toContain('no-scroll');
        });
    });

    describe('componentWillUnmount', () => {
        test('removes event listeners from window', () => {
            window.removeEventListener = jest.fn();
            const wrapper = shallow(<Modal />);
            wrapper.instance().componentWillUnmount();
            expect(window.removeEventListener).toHaveBeenCalledTimes(2);
            expect(window.removeEventListener).toHaveBeenCalledWith('keyup', wrapper.instance().handleCloseKeyUp);
            expect(window.removeEventListener).toHaveBeenCalledWith('focus', wrapper.instance().handleFocus, true);
        });

        test('removes "no-scroll" class from body', () => {
            document.body.classList.add('no-scroll');
            expect(document.body.classList).toContain('no-scroll');
            const wrapper = shallow(<Modal />);
            wrapper.instance().componentWillUnmount();
            expect(document.body.classList).not.toContain('no-scroll');
        });
    });

    describe('handleFocus', () => {
        test('focuses on the close button', () => {
            const wrapper = shallow(<Modal />);
            const containerRefMock = {
                current: {
                    focus: jest.fn(),
                    contains() {
                        return false;
                    },
                },
            };
            const closeButtonRefMock = {
                current: {
                    focus: jest.fn(),
                },
            };
            wrapper.instance().container = containerRefMock;
            wrapper.instance().closeButton = closeButtonRefMock;
            wrapper.instance().handleFocus({
                currentTarget: {},
            });
            expect(closeButtonRefMock.current.focus).toHaveBeenCalledTimes(1);
        });

        test('does not focus on the close button', () => {
            const wrapper = shallow(<Modal />);
            const containerRefMock = {
                current: {
                    focus: jest.fn(),
                    contains() {
                        return true;
                    },
                },
            };
            const closeButtonRefMock = {
                current: {
                    focus: jest.fn(),
                },
            };
            wrapper.instance().container = containerRefMock;
            wrapper.instance().closeButton = closeButtonRefMock;
            wrapper.instance().handleFocus({
                currentTarget: {},
            });
            expect(closeButtonRefMock.current.focus).not.toHaveBeenCalled();
        });
    });
});
