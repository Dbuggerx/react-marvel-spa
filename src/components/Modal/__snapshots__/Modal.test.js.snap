// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal render renders properly 1`] = `
<div
  className="modal modal--opened"
>
  <dialog
    aria-describedby="modal-content"
    aria-labelledby="modal-title"
    aria-modal="true"
    className="modal__container"
    open={true}
  >
    <button
      aria-label="Close"
      className="modal__close"
      onKeyUp={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        aria-hidden="true"
      >
        Ã—
      </span>
    </button>
    <div
      className="modal__title"
      id="modal-title"
    >
      Testing
    </div>
    <div
      id="modal-content"
    >
      Test
    </div>
  </dialog>
</div>
`;
